distributed/sharded_tensor/tensor_parallel.py

```diff
@@ -0,0 +1,113 @@
+import os
+import tempfile
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import torch.nn as nn
+	
+from torch.distributed._shard import shard_module
+from torch.distributed._shard.sharded_optim import (
+    ShardedOptimizer,
+    named_params_with_sharded_tensor,
+)
+from torch.distributed._shard.sharding_plan import ShardingPlan
+from torch.distributed._shard.sharding_spec import ChunkShardingSpec
+
+"""
+This is script to test Tensor Parallel(TP) on a toy model in a 
+Megetron-LM SPMD style. We show a E2E working flow from forward,
+backward and optimization.
+"""
+
+
+def setup(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12355'
+
+    # initialize the process group
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+class ToyModel(nn.Module):
+    def __init__(self):
+        super(ToyModel, self).__init__()
+        self.net1 = nn.Linear(10, 32)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(32, 5)
+
+    def forward(self, x):
+        return self.net2(self.relu(self.net1(x)))
+
+	
+def _generate_chunk_sharding_spec(world_size):
+    placements = [f"rank:{idx}/cuda:{idx}" for idx in range(world_size)]
+    colwise_spec = ChunkShardingSpec(
+        dim=0,
+        placements=placements,
+    )
+    rowwise_spec = ChunkShardingSpec(
+        dim=1,
+        placements=placements,
+    )
+    return colwise_spec, rowwise_spec
+
+
+def _get_toy_module_optim(module, lr):
+    return ShardedOptimizer(
+        dict(named_params_with_sharded_tensor(module)),
+        torch.optim.SGD,
+        lr=lr,
+    )
+
+
+def _get_toy_module_sharding_plan(specs):
+    colwise_spec, rowwise_spec = specs[0], specs[1]
+    return ShardingPlan(
+        plan={
+            "net1.weight": colwise_spec,
+            "net2.weight": rowwise_spec,
+        },
+        output_plan={
+            "net2": colwise_spec,
+        },
+        return_local_tensor=["net2"],
+    )
+
+
+def demo_tp(rank, world_size):
```

Comment:
Think of this as a tutorial/example in terms of a PyTorch user with no background of our sharding APIs coming across this. As a result, lets add detailed comments in each section explaining exactly what is happening in each significant line.